### PR TITLE
Fix segfault from requests_cache

### LINKED ISSUE #196: Enable requests caching only for a subset of URLs
Now requests caching is enabled globally, this is quite bad because any import of aiidalab package might result in a strange behaviour of other tools that rely on `requests` package.

Unfortunately, I do not see a way to enable the caching only for specific URLs. 

**UPD**:
Using [`requests_cache.CachedSession`](https://requests-cache.readthedocs.io/en/stable/user_guide.html#sessions) is the way to go.

### LINKED ISSUE #413: Test flake
One of the tests seems to fail randomly quite often

```python
FAILED tests/test_appclass.py::test_dependencies - traitlets.traitlets.TraitError: Version v22.11.1 is not available for quantum-espresso app.
```

CC @unkcpz

### LINKED ISSUE #446: Segfault in pytest for 3.11/3.12
So the Python 3.11/3.12 test started failing with a ... Segfault? :scream:

https://github.com/aiidalab/aiidalab/actions/runs/10298925056/job/28505124693

```
Fatal Python error: Segmentation fault

Thread 0x00007fc2fc48f640 (most recent call first):
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests_cache/backends/sqlite.py", line 278 in close
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests_cache/backends/base.py", line 116 in close
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests_cache/session.py", line 342 in close
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests/sessions.py", line 455 in __exit__
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests/api.py", line 58 in request
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests/api.py", line 73 in get
  File "/home/runner/work/aiidalab/aiidalab/aiidalab/utils.py", line 62 in load_app_registry_index
  File "/home/runner/work/aiidalab/aiidalab/aiidalab/app.py", line 138 in is_registered
  File "/home/runner/work/aiidalab/aiidalab/aiidalab/app.py", line 193 in available_versions
  File "/home/runner/work/aiidalab/aiidalab/aiidalab/app.py", line 910 in _refresh_versions
  File "/home/runner/work/aiidalab/aiidalab/aiidalab/app.py", line 924 in refresh
  File "/home/runner/work/aiidalab/aiidalab/aiidalab/utils.py", line 148 in wrapped
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/threading.py", line 982 in run
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/threading.py", line 1045 in _bootstrap_inner
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/threading.py", line 1002 in _bootstrap

Current thread 0x00007fc2ffdc6b80 (most recent call first):
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests_cache/backends/sqlite.py", line 312 in __getitem__
  File "<frozen _collections_abc>", line 774 in get
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests_cache/backends/base.py", line 77 in get_response
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests_cache/session.py", line 218 in send
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests/sessions.py", line 589 in request
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests_cache/session.py", line 183 in request
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests/api.py", line 59 in request
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/requests/api.py", line 73 in get
  File "/home/runner/work/aiidalab/aiidalab/aiidalab/utils.py", line 62 in load_app_registry_index
  File "/home/runner/work/aiidalab/aiidalab/aiidalab/app.py", line 138 in is_registered
  File "/home/runner/work/aiidalab/aiidalab/aiidalab/app.py", line 323 in find_incompatibilities
  File "/home/runner/work/aiidalab/aiidalab/aiidalab/app.py", line 882 in _is_compatible
  File "/home/runner/work/aiidalab/aiidalab/aiidalab/app.py", line 927 in refresh
  File "/home/runner/work/aiidalab/aiidalab/aiidalab/utils.py", line 148 in wrapped
  File "/home/runner/work/aiidalab/aiidalab/tests/test_appclass.py", line 16 in test_init_refresh
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/python.py", line 162 in pytest_pyfunc_call
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/python.py", line 1632 in runtest
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/runner.py", line 173 in pytest_runtest_call
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/runner.py", line 241 in <lambda>
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/runner.py", line 341 in from_call
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/runner.py", line 240 in call_and_report
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/runner.py", line 135 in runtestprotocol
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/runner.py", line 116 in pytest_runtest_protocol
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/main.py", line 364 in pytest_runtestloop
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/main.py", line 339 in _main
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/main.py", line 285 in wrap_session
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/main.py", line 332 in pytest_cmdline_main
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/config/__init__.py", line 178 in main
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/_pytest/config/__init__.py", line 206 in console_main
  File "/opt/hostedtoolcache/Python/3.11.9/x64/bin/pytest", line 8 in <module>
```

### PR DESCRIPTION
**Pull request by danielhollas**
As #196 explains, we now use a global cache for *all* `requests` calls! That's quite scary since it influences apps that import `aiidalab` package, such as AWB or home app! 

Here we fix this using the approach suggested in https://github.com/aiidalab/aiidalab/issues/196#issuecomment-899196469 Moreover, this seems to fix the segfault that we started seeing with `requests_cache>=1.0.0` described in #446.

Fixes #446 ( :crossed_fingers: ). Fixes #196 
Hopefully also fixes #413, I've removed the call to the app registry from tests.
 
### Testing 

Because we generally want to avoid fetching app index over network in tests, I've tested manually.

1. Launch a container with the latest `ghcr.io/aiidalab/full-stack:edge` image
2. Navigate to the app store page and observe that it crashes with "Kernel died" error, this is cause by the segfault described in #446 
3. Setup a dev environment with the help of aiidalab-dev
```
pip install git+https://github.com/aiidalab/aiidalab-dev
develop-aiidalab setup
```
4. Checkout this branch
```
cd local/aiidalab && git fetch origin fix-segfault && git switch fix-segfault
```
5. Now refresh the appstore page, it should now load.

Here's a script where you can reproduce the segfault from command line

```python
import logging
from home.app_store import AiidaLabAppStore    
      
logging.basicConfig(level='INFO')    
logging.getLogger('requests_cache').setLevel('DEBUG')    
      
store = AiidaLabAppStore()
```

TODO: Might be able to write a unit test with local file with either:
- https://pypi.org/project/requests-file/
- https://requests-mock.readthedocs.io/en/latest/index.html
- https://github.com/kiwicom/pytest-recording

### Debugging

Add logging calls to `aiidalab.utils.load_app_registry_index` or `load_app_registry_entry`:
```python
        response = _session.get(f"{AIIDALAB_REGISTRY}/apps_index.json")
        import time
        logger.info(f"Time: {time.ctime(int(time.time()))} Used cache: {response.from_cache}")
        return response.json()
```

### CODE REVIEW

File: tests/conftest.py
Comment (danielhollas) [inline tests/conftest.py:50]: `is_registered` method fetches the app index, which we want to avoid in tests. This was likely the cause of the test flakiness that we saw in #413.

File: tests/conftest.py
Comment (danielhollas) [inline tests/conftest.py:14]: I've tried this but this approach does not work for Python < 3.12, because `importlib.resources.files()` requires a package name (not a module) and `tests/` is not a package (doesn't have `__init__.py` file).

https://docs.python.org/3/library/importlib.resources.html#importlib.resources.files

File: aiidalab/utils.py
Comment (danielhollas) [inline aiidalab/utils.py:36]: Per  https://requests-cache.readthedocs.io/en/stable/user_guide/general.html#sessions

The old global approach is described here:
https://requests-cache.readthedocs.io/en/stable/user_guide/general.html#patching

and it's downsides, which specifically mention multithreaded use: https://requests-cache.readthedocs.io/en/stable/user_guide/general.html#patching-limitations-potential-issues

**Comment by danielhollas** [comment]
@mikibonacci this should fix the app store crash that you reported in https://github.com/aiidalab/aiidalab-docker-stack/pull/482. Would be great if you could give this a look and test, thanks!

File: tests/conftest.py
Comment (danielhollas) [reply tests/conftest.py:14]: (unrelated to this PR)

**Comment by mikibonacci** [comment]
Hi @danielhollas , thank you for the work! I am on holidays until next week, I will test as soon as I go back to work!

**Comment by danielhollas** [comment]
@unkcpz will you have time to review this and other PRs on this repo this week? This is currently blocking the release of image with aiida 2.6. 

Relatedly, is there somebody else I should be asking for review for this repo? Would be good to have at least three active maintainers. CC @yakutovicha

**Comment by unkcpz** [comment]
Hi @danielhollas, I think I don't have time at the moment, sorry. Had quite some scientific things need to deliver in two weeks and I'll take some holiday after.

**Comment by danielhollas** [comment]
> Hi @danielhollas, I think I don't have time at the moment, sorry. Had quite some scientific things need to deliver in two weeks and I'll take some holiday after.

Absolutely no problem, this is why I am asking. 

I would like to get unblocked on this. Is there somebody else you think could help maintaining this repo? Or should I go ahead and merge the fixes and deal with review later? Essentially to me it's not great to be constantly block on review even on relatively small bugfix PRs and it's not fair that you're currently the only one that reviews my stuff to the core repositories.

**Comment by superstar54** [comment]
I can do it. I will work on aiidalab this week.

File: tests/conftest.py
Comment (superstar54) [inline tests/conftest.py:38]: Why is the import moved here?

File: aiidalab/utils.py
Comment (superstar54) [inline aiidalab/utils.py:35]: Does this problem not exist in Quantum Mobile anymore?

**Review by superstar54 — CHANGES_REQUESTED** [review]
Hi @danielhollas , thanks for the work. I added some small change requests.

File: aiidalab/utils.py
Comment (danielhollas) [reply aiidalab/utils.py:35]: This code is 3 years old, so I would certainly hope so. :-) Would you be able to check with the latest version of Quantum Mobile? If that is too much work, I can put this back if it is of concern. I don't actually know how QM is build and how aiidalab is integrated into it, but I would hope that the dependencies of this package are correctly installed.

File: tests/conftest.py
Comment (danielhollas) [reply tests/conftest.py:38]: I don't remember if I had a big reason for this TBH. This is test code so I don't care much either way.

File: aiidalab/utils.py
Comment (superstar54) [reply aiidalab/utils.py:35]: I think @unkcpz works on the latest QM. Maybe he can comment when he gets time. For now, I would suggest putting it back.

Comment (danielhollas) [reply aiidalab/utils.py]: Okay, I added it back and opened https://github.com/aiidalab/aiidalab/issues/452 as a follow-up.

Unlike in the old code, the code path that does not use requests_cache uses the `requests.Session` object, which should give much better performance as well since the HTTP connection should be preserved between calls.

**Review by superstar54 — APPROVED** [review]
LGTM!

**Pull request merged**
